compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1977,11 +1994,13 @@ class Typer extends Namer
    */
   def annotContext(mdef: untpd.Tree, sym: Symbol)(using Context): Context = {
     def isInner(owner: Symbol) = owner == sym || sym.is(Param) && owner == sym.owner
-    val c = ctx.outersIterator.dropWhile(c => isInner(c.owner)).next()
-    c.property(ExprOwner) match {
-      case Some(exprOwner) if c.owner.isClass => c.exprContext(mdef, exprOwner)
-      case _ => c
+    val outer = ctx.outersIterator.dropWhile(c => isInner(c.owner)).next()
+    val c = outer.property(ExprOwner) match {
+      case Some(exprOwner) if outer.owner.isClass => outer.exprContext(mdef, exprOwner)
+      case _ => outer
     }
+    if (c.isJava) c.fresh.setProperty(JavaAnnotationArg, ())
+    else c
```

Comment:
My logic here was that `JavaAnnotationArg` implies that `c.isJava` is true and that we are currently typechecking annotation arguments.
If it is guaranteed that `isJava` being true inside of `typedNamedArg` implies that we are typechecking annotation arguments, then indeed it may be redundant. But I'm not sure if it is guaranteed now and it stays guaranteed in the future.

Reply:
> But I'm not sure if it is guaranteed now and it stays guaranteed in the future.

I think it's safe to assume this if you leave a comment explaining the assumption in the code, we can always adapt the code if Java somehow changes, but I wouldn't worry too much about it.